### ISSUE TITLE
Audio stream sometimes failing to start with steam-launched games

### ISSUE THREAD

**Issue by mtheilm** [colinmarc/magic-mirror#3]
## Environment
System: Apple M1, MacBook Air
OS: 14.4.1
magic-mirror release: [mmclient-v0.1.0](https://github.com/colinmarc/magic-mirror/releases/tag/mmclient-v0.1.0)

## Client
Small nit: Video on client times out when first starting magic-mirror while waiting for Steam to go through startup process. Waiting ~30s and re-attempting connects just fine. 

Bug: no audio coming through when running magic-mirror. Command run and output as follows:

```
% ./mmclient "<IP>:9599" steam --resolution 1080 --codec h265
2024-04-27T19:39:39.158518Z  WARN mm_client::conn: skipping TLS verification for private server address
2024-04-27T19:39:39.329225Z  INFO mmclient: launching a new session for for app "steam"
2024-04-27T19:39:39.329390Z  INFO mmclient: launching session for app "steam"
2024-04-27T19:39:41.076812Z  INFO mm_client::vulkan: selected gpu: "Apple M1" (0)
2024-04-27T19:39:41.080826Z  WARN mm_client::vulkan: no vulkan video support found!
2024-04-27T19:39:41.099697Z  WARN mm_client::render: present mode MAILBOX not available, using FIFO (available: [FIFO, IMMEDIATE])
2024-04-27T19:39:41.139711Z  INFO mm_client::audio: using audio output device: MacBook Air Speakers
2024-04-27T19:39:41.173763Z  INFO mmclient: successfully (re)attached to session msg.session_id=448129
2024-04-27T19:39:41.340616Z  WARN mm_client::render: present mode MAILBOX not available, using FIFO (available: [FIFO, IMMEDIATE])
2024-04-27T19:39:51.515735Z ERROR mmclient: timed out waiting for video frames
% ./mmclient "<IP>:9599" steam --resolution 1080 --codec h265
2024-04-27T19:41:24.672610Z  WARN mm_client::conn: skipping TLS verification for private server address
2024-04-27T19:41:24.863096Z  INFO mm_client::vulkan: selected gpu: "Apple M1" (0)
2024-04-27T19:41:24.866131Z  WARN mm_client::vulkan: no vulkan video support found!
2024-04-27T19:41:24.878073Z  WARN mm_client::render: present mode MAILBOX not available, using FIFO (available: [FIFO, IMMEDIATE])
2024-04-27T19:41:24.915649Z  INFO mm_client::audio: using audio output device: MacBook Air Speakers
2024-04-27T19:41:24.944628Z  INFO mmclient: successfully (re)attached to session msg.session_id=448129
2024-04-27T19:41:25.091325Z  WARN mm_client::render: present mode MAILBOX not available, using FIFO (available: [FIFO, IMMEDIATE])
2024-04-27T19:41:26.782583Z ERROR mm_client::conn: protocol error: invalid message type: 1044
2024-04-27T19:41:34.294174Z ERROR mm_client::conn: protocol error: protobuf decode error: failed to decode Protobuf message: invalid varint
```

## Server
Linux version: 6.6.26-1-MANJARO
GPU: Nvidia P1000
```
./mmserver --bind "<IP>:9599" -i ~/src/magic-mirror/steam.toml --bug-report
```

Generated bug report attached.
[mmserver.log](https://github.com/colinmarc/magic-mirror/files/15139628/mmserver.log)

**Comment by colinmarc** [comment]
Oh hi Mark!

Can you run the latest server release as well? I can tell it's an old version of the server because it doesn't have this log line: https://github.com/colinmarc/magic-mirror/blob/c1fbaf1a7ef5f01939729d00c27ebc70b4699d8d/mm-server/src/main.rs#L71

The binary release (v0.1.0) should be fine. Unfortunately, that probably won't fix the issue. I'm suspicious of the 30s delay starting steam being the root cause somehow. That shouldn't really be happening.

**Comment by colinmarc** [comment]
Can you try running the following and seeing if it makes steam start up faster?

    systemctl start --user xdg-desktop-portal

If that works, and you're on a headless server, you can disable it on startup with:

    systemctl disable --user xdg-desktop-portal

**Comment by mtheilm** [comment]
The above bug report is running fresh from the binary v0.1.0! Re-downloading from https://github.com/colinmarc/magic-mirror/releases/download/mmserver-v0.1.0/mmserver-v0.1.0-linux-amd64.tar.gz still doesn't yield `starting up`. Also rebuilding from the `main` branch doesn't show that debug note either.

 And secondly:
```
 systemctl stop --user xdg-desktop-portal
Failed to stop xdg-desktop-portal.service: Unit xdg-desktop-portal.service not loaded.
```
so I don't think that's slowing steam down.

**Comment by colinmarc** [comment]
> The above bug report is running fresh from the binary v0.1.0! Re-downloading from https://github.com/colinmarc/magic-mirror/releases/download/mmserver-v0.1.0/mmserver-v0.1.0-linux-amd64.tar.gz still doesn't yield `starting up`. Also rebuilding from the `main` branch doesn't show that debug note either.

My bad: https://github.com/colinmarc/magic-mirror/commit/f708ad2d8e5ddc9fb17ac023fef8f81706c31be7

Steam is just not creating an audio stream. I'm not sure why. Can you post a log where you also launch a game via steam and click around in the menu screen?

I also find this plain weird, I've never gotten an error like this and I don't understand how it would happen:

```
2024-04-27T19:41:26.782583Z ERROR mm_client::conn: protocol error: invalid message type: 1044
2024-04-27T19:41:34.294174Z ERROR mm_client::conn: protocol error: protobuf decode error: failed to decode Protobuf message: invalid varint
```

**Comment by colinmarc** [comment]
In order to isolate the slow start thing, can you try this patch on the client side?

```diff
diff --git a/mm-client/src/bin/mmclient.rs b/mm-client/src/bin/mmclient.rs
index 101c3a7..f0fce32 100644
--- a/mm-client/src/bin/mmclient.rs
+++ b/mm-client/src/bin/mmclient.rs
@@ -40,7 +40,7 @@ use mm_client::video;
 use mm_client::video::VideoStreamEvent;
 use mm_client::vulkan;

-const INIT_TIMEOUT: time::Duration = time::Duration::from_secs(10);
+const INIT_TIMEOUT: time::Duration = time::Duration::from_secs(60);
 const MAX_FRAME_TIME: time::Duration = time::Duration::from_nanos(1_000_000_000 / 24);
 const RESIZE_COOLDOWN: time::Duration = time::Duration::from_millis(500);
```

And then just wait until it fully starts.

**Comment by mtheilm** [comment]
Ok some new data: 
1. upping the timeout successfully let's steam launch on the first try! 
2. running with h264 works; running with h265 does not work
3. logs from running Into the Breach and clicking around a bit in-menu attached

server (still not showing `starting up` fwiw), on `f708ad2d8e5ddc9fb17ac023fef8f81706c31be7`:
```
2024-04-27T16:46:32.493543Z  WARN mmserver: generating bug report files in: "/tmp/magic-mirror-6e5e975e-0308-4bfe-a626-8e4b6b9a33c0"
2024-04-27T16:46:32.544090Z  INFO mmserver::vulkan: selected gpu: "Quadro P1000" (0)
2024-04-27T16:46:32.567220Z  INFO mmserver::vulkan: hardware encoding support: (h264: true, h265: true, av1: false)
2024-04-27T16:46:32.568367Z  INFO mmserver: listening on 100.124.198.118:9599
2024-04-27T16:46:36.193948Z  INFO mmserver::session: launching session session_id=888230 application="steam"
2024-04-27T16:46:37.423185Z  INFO mmserver::session: new attachment session_id=888230 attachment_id=967571 operator=true
^C2024-04-27T16:47:27.451884Z  INFO mmserver: bug report files saved to: "/tmp/magic-mirror-6e5e975e-0308-4bfe-a626-8e4b6b9a33c0"

```
[logs.tar.gz](https://github.com/colinmarc/magic-mirror/files/15140076/logs.tar.gz)

**Comment by colinmarc** [comment]
> running with h264 works; running with h265 does not work

Audio works with h264 but fails with h265? That's very strange indeed.

**Comment by mtheilm** [comment]
Oh, apologies: that was my mistake: audio works with neither. No video issues when launching from the beginning with h264 or h265, I had tried to re attach with a different codec.

**Comment by colinmarc** [comment]
Heh, reattaching with a different codec should also work. I'll try to reproduce that.

Can you try big picture mode? You can just change `command` to `["steam", "-gamepadui"]` in your app config. I generally test with that and it makes a big sound right at the beginning, so that might help us isolate the issue.

**Comment by colinmarc** [comment]
Per DMs:

 - Big picture works for audio, and usually in-game as well, but it sometimes cuts out
 - The server is seeing CPU more or less completely pegged by the game (ITB - software rendering?)

I think the bug is probably something where the pulse server isn't getting scheduled and then gets stuck.

> Heh, reattaching with a different codec should also work. I'll try to reproduce that.

I haven't been able to reproduce this.

**Comment by colinmarc** [comment]
I reproduced the steam hang on a cloud box, but it went away if I use the following in my config:

```
environment = { "__GLX_VENDOR_LIBRARY_NAME" = "nvidia" }
```

I think your steam/ITB are actually not using the GPU, which is pegging CPU and causing other problems. Could still be a bug in the pulse server, though.

**Comment by colinmarc** [comment]
Pretty sure this should be fixed in 0.2.0. Let me know if it isn't!
